_posts/2025-05-14-#32317.md

```diff
@@ -7,32 +7,68 @@ authors: [TheCharlatan]
+approach rather than a complete rewrite, its current interface is still strongly influenced by
+Bitcoin Core's architecture, requirements and assumptions.
 
+The UTXO set is a crucial component of Bitcoin Core's architecture, but it is an implementation
+detail, and one that other Bitcoin node implementations may choose not to implement. For example:
+- [Utreexo](https://bitcoinops.org/en/topics/utreexo/) nodes, such as
+  [Floresta](https://github.com/vinteumorg/Floresta) rely on an accumulator instead of a UTXO set
+- [SwiftSync](https://gist.github.com/RubenSomsen/a61a37d14182ccd78760e477c78133cd) is
+  near-stateless, and does not have the concept of a UTXO set.
+
+The motivation behind this PR is to, in future work, allow kernel users without a UTXO set to
+validate a transaction by providing the validation function the specific UTXTOs it is spending.
 
 ## Questions
 
-1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
+   NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your
+   review approach?
+
+2. Why is carving out the new `SpendBlock()` function from `ConnectBlock()` helpful for this PR? How
+   would you compare the purpose of the two functions?
+
+4. Do you see another benefit of this decoupling, besides allowing kernel usage without a UTXO set?
+
+5. Especially during IBD, transaction validation must be fast. Are there any changes where you have
+   concerns about performance? Which measures can you identify this PR has adopted to minimize the
+   performance impact of this refactor?
+
+6. `SpendBlock()` takes a `CBlock block`, `CBlockIndex pindex` and `uint256 block_hash` parameter,
+   all referencing the block being spent. Why do we need 3 parameters to do that?
+
+6. `CCoinsViewCache` has two methods `AccessCoin()` and `GetCoin`() that both return a `Coin`-like
+   type. What is the difference between both methods, and when should which be used?
+
+7. The first commits in this PR refactor `CCoinsViewCache` out of the function signature of a couple
+   of validation functions. Does `CCoinsViewCache` hold the entire UTXO set? Why is that (not) a
+   problem? Does this PR change that behaviour?
+
+8. Why does [commit
+   `a7e4132`](https://github.com/bitcoin-core-review-club/bitcoin/commit/a7e41326234d3a381fdde0924af74c6561b10798)
+   use explicit template instantiation for `GetP2SHSigOpCount()`?
+
+9. Following on the previous question, why is there explicit instantiation of the
+   `<std::span<std::reference_wrapper<const Coin>>>` template when a `std::reference_wrapper<T>`
+   already [implicitly
+   converts](https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper#:~:text=but%20they%20are,T%26)
+   to `T&`?
```

Comment:
I actually need to dig into this reason again, but msvc was giving me a lot of pain when I tried to implicitly convert it.

Reply:
Hmm, the answer I had in mind was that implicit conversion does not work for container value-types, because the reference wrappers are different objects allocated on the heap? (Although I think you might be thinking about the `GetCoin()` helper, which does not involve a container, and indeed confused me (and I have as an outstanding review comment) as to why we need it).